challenger.go

```diff
@@ -79,3 +234,104 @@ func (l *LndChallenger) NewChallenge(price int64) (string, lntypes.Hash, error)
 
 	return response.PaymentRequest, paymentHash, nil
 }
+
+// VerifyInvoiceStatus checks that an invoice identified by a payment
+// hash has the desired status. To make sure we don't fail while the
+// invoice update is still on its way, we try several times until either
+// the desired status is set or the given timeout is reached.
+//
+// NOTE: This is part of the auth.InvoiceChecker interface.
+func (l *LndChallenger) VerifyInvoiceStatus(hash lntypes.Hash,
+	state lnrpc.Invoice_InvoiceState, timeout time.Duration) error {
+
+	// Prevent the challenger to be shut down while we're still waiting for
+	// status updates.
+	l.wg.Add(1)
+	defer l.wg.Done()
+
+	var (
+		condWg         sync.WaitGroup
+		doneChan       = make(chan struct{})
+		timeoutReached bool
+		hasInvoice     bool
+		invoiceState   lnrpc.Invoice_InvoiceState
+	)
+
+	// First of all, spawn a goroutine that will signal us on timeout.
+	// Otherwise if a client subscribes to an update on an invoice that
+	// never arrives, and there is no other activity, it would block
+	// forever in the condition.
+	condWg.Add(1)
+	go func() {
+		defer condWg.Done()
+
+		select {
+		case <-doneChan:
+		case <-time.After(timeout):
+		case <-l.quit:
+		}
+
+		l.invoicesCond.L.Lock()
+		timeoutReached = true
+		l.invoicesCond.Broadcast()
+		l.invoicesCond.L.Unlock()
+	}()
+
+	// Now create the main goroutine that blocks until an update is received
+	// on the condition.
+	condWg.Add(1)
+	go func() {
+		defer condWg.Done()
+		l.invoicesCond.L.Lock()
+
+		// Block here until our condition is met or the allowed time is
+		// up. The Wait() will return whenever a signal is broadcast.
+		invoiceState, hasInvoice = l.invoiceStates[hash]
+		for !(hasInvoice && invoiceState == state) && !timeoutReached {
+			l.invoicesCond.Wait()
+
```

Comment:
Check the quit signal here to make sure the goroutine is able to exit when we need to do an early shutdown, but the invoice isn't yet known/settled?

Reply:
The goroutine above would exit on `l.quit` and set `timeoutReached` to `true` and also broadcast an update on the `Cond` object. That would cause the `Wait()` to return and we wouldn't enter the loop again afterward because `timeoutReached` is now true. Or what am I missing?